staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,1059 @@
+	ca.vniToVnStateMutex.Lock()
+	defer ca.vniToVnStateMutex.Unlock()
+	vnState := ca.vniToVnState[vni]
+	if vnState != nil {
+		delete(vnState.localAtts, localAttName)
+		if len(vnState.localAtts) == 0 {
+			delete(ca.vniToVnState, vni)
+			return vnState
+		}
+	}
+	return nil
+}
+
+func (ca *ConnectionAgent) stopRemAttInformerAndEnqueueRemAttRefs(state *vnState, vni uint32) {
+	close(state.remoteAttsInformerStopCh)
+	for aRemoteAttName := range state.remoteAtts {
+		ca.queue.Add(attQueueRef{
+			vni:   vni,
+			local: false,
+			nsn: k8stypes.NamespacedName{
+				Namespace: state.namespace,
+				Name:      aRemoteAttName,
+			},
+		})
+	}
+}
+
+func (ca *ConnectionAgent) removeRemoteAttFromVNState(remAttName string, vni uint32) {
+	ca.vniToVnStateMutex.Lock()
+	defer ca.vniToVnStateMutex.Unlock()
+	vnState := ca.vniToVnState[vni]
+	if vnState != nil {
+		delete(vnState.remoteAtts, remAttName)
+	}
+	return
+}
+
+func (ca *ConnectionAgent) getRemoteAttListerForVNI(vni uint32) koslisterv1a1.NetworkAttachmentNamespaceLister {
+	ca.vniToVnStateMutex.RLock()
+	defer ca.vniToVnStateMutex.RUnlock()
+	vnState := ca.vniToVnState[vni]
+	if vnState == nil {
+		return nil
+	}
+	return vnState.remoteAttsLister
+}
+
+func (ca *ConnectionAgent) setIfcNameAndHostIPInLocalAttStatusIfNeeded(localAtt *netv1a1.NetworkAttachment,
+	ifcName string,
+	ifcNameMightNeedUpdate bool) (updatedAtt *netv1a1.NetworkAttachment, err error) {
+
+	if ca.localAttNeedsStatusUpdate(localAtt, ifcName, ifcNameMightNeedUpdate) {
+		updatedAtt, err = ca.setIfcNameAndHostIPInLocalAttStatus(localAtt, ifcName, ifcNameMightNeedUpdate)
+	}
+	return
+}
+
+func (ca *ConnectionAgent) localAttNeedsStatusUpdate(localAtt *netv1a1.NetworkAttachment,
+	ifcName string,
+	ifcNameMightNeedUpdate bool) bool {
+
+	return !ca.hostIP.Equal(gonet.ParseIP(localAtt.Status.HostIP)) ||
+		(ifcNameMightNeedUpdate && localAtt.Status.IfcName != ifcName)
+}
+
+func (ca *ConnectionAgent) setIfcNameAndHostIPInLocalAttStatus(localAtt *netv1a1.NetworkAttachment,
+	ifcName string,
+	ifcNameMightNeedUpdate bool) (updatedAtt *netv1a1.NetworkAttachment, err error) {
+
+	att2 := localAtt.DeepCopy()
+	att2.Status.HostIP = ca.hostIP.String()
+	if ifcNameMightNeedUpdate {
+		att2.Status.IfcName = ifcName
+	}
+	updatedAtt, err = ca.netv1a1Ifc.NetworkAttachments(att2.Namespace).Update(att2)
+	return
+}
+
+func (ca *ConnectionAgent) clearHostIPAndIfcNameFromLocalAttStatus(localAttNSN k8stypes.NamespacedName, ifcName string) (err error) {
+	patchData := prepareStatusPatchData(ca.hostIP, ifcName)
+	_, err = ca.netv1a1Ifc.NetworkAttachments(localAttNSN.Namespace).Patch(localAttNSN.Name, k8stypes.JSONPatchType, patchData, statusSubRes)
+	return
+}
+
+func prepareStatusPatchData(hostIP gonet.IP, ifcName string) []byte {
+	return []byte(fmt.Sprintf(`[{"op": "test", "path": "/status/hostIP", "value": "%s"},
+	{"op": "remove", "path": "/status/hostIP"},
+	{"op": "test", "path": "/status/ifcName", "value": "%s"},
```

Comment:
I suspect we want both tests to come first.
Also, shouldn't we also test the VNI?